### DISCUSSION TITLE
Complete Noob with Questions (a bit lengthy, my apologies)

### DISCUSSION THREAD

**Discussion by COBRABYTE** [Homebrew/discussions#737] [Getting started]
Hey everyone, I'm completely new to programming and even using the terminal (on Mac), and I need some help. I have been trying to do as much research as I can to prepare for learning and doing programming. One of the things that I read was to install Homebrew, as it was stated that it would become much more beneficial to me down the line having an environment with Homebrew and it's packages installed and all ready to go.

Before I continue, I want to mention a couple of things as it may help you diagnose the issues I'm having. First and foremost, as I said I am completely new to programming and the terminal. In fact, I have only just written my first hello_world program in Python. Likewise, I have never used the terminal outside of recently learning some basic commands for navigating the Mac's folder structure. Additionally, I have also installed oh-my-zsh (per reading instructions on how to do it; I had no idea what I was doing as I was doing it or really what all of the stuff in the terminal meant). I should note that the real reason for installing it was to hopefully have a more customizable terminal with different colors for different things (this is not so much for aesthetic purposes as it is for my visual needs; I am visually impaired and having different colors for different types of things helps me to distinguish between them much more efficiently, however, it doesn't seem to have as much color differentiation as I had initially anticipated). Anyways, so that is where I'm at at the moment:

- completely new to programming (Python for now) and terminal
- installed oh-my-zsh

Here are my issues:

I installed Homebrew just fine--no errors that I saw. Then I read that I should install cask. After having read some basic commands, I though it wise to use this: $ brew info cask so that I could find out about cask. At that moment, my assumption was that I had to work backwards installing the dependencies (track backwards to find dependencies that needed no dependencies and install those first). That led me to first do $ brew install pkg-config. When I did this I had gotten an error: 

1. "Failed to read Mach-O binary" and then an address to a directory or maybe a file 

 Unfortunately, I did not write it down or save the text to a text file (which I should have done).

I later found out that you can just install something that requires dependencies, and it will install all needed dependencies. So I just went straight for $ brew install cask. When I did this, I got a couple errors (this time I saved them):

1. Error: Failed to read Mach-O binary: /usr/local/Cellar/jemalloc/5.2.1_1/lib/pkgconfig/jemalloc.pc

2. Error: Your CLT does not support macOS 11.
It is either outdated or was modified.
Please update your CLT or delete it if no updates are available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

3. Error: An exception occurred within a child process:
  SystemExit: exit
    FAIL: 1

So now for the questions:

1. What are these errors?
2. Do I need to fix them? (stupid question, but I mean do I manually need to fix them or are they self fixing)
3. How would I fix these errors?
4. In the future, if I want to see if there are any issues/errors with Homebrew or any of its packages that I have installed, is there a way to check for them in the terminal (or anywhere else for that matter)?

bonus questions:

5. Could someone help me to customize my terminal so that it has lots of different colors (and/or distinguishing font characteristics i.e. bold, underlined, italic, and so forth, for different things?

 For example, I'd like to have different colors to distinguish between directories, files, hidden files and such. I really don't know what other types of things I will encounter in the terminal, but I would just like to have as much differentiation as I can, as that will make it easier for me visually to be more efficient in the terminal.

6. If I wanted to learn and understand the terminal very well, be it to customize it or to use it, are there any books or resources available that would be for absolute beginners? If so, do you have any suggestions on specific books or resources?

I know that I can just Google something about the terminal as I go, but I'd like to find something that allows me to build a foundation, that has a finite starting point and continues to guide me down a path, building on top of that foundation. When I Google something specific about a terminal task, and I read an article or post, I end up just coming away with more questions and no real understanding.

Ok, so that's it. I'm really sorry for the length of this post. I'm just completely new, have no idea what I'm doing, and have tons of questions. Any help would be greatly appreciated.

**Comment by Rylan12** [comment]
Hi there,

I'll try my best to help you out a little. Just so you know, this discussion forum is specifically geared toward Homebrew related things, so your "bonus questions" may be best suited for other places.

---

Based on the error messages I'm seeing, it appears that your command-line tools (CLT) are out of date. There are a few ways to solve this. The easiest is probably to check is System Preferences → Software Update to see if there are any updates (especially ones for the CLT). Try installing those updates.

If that doesn't work, you can download the latest version of the CLT from the [Apple Developer website](https://developer.apple.com/download/more/). You may need to make a free Apple Developer account. You're looking for the latest stable version of the "Command Line Tools for Xcode". Currently, it looks like the latest stable version is "Command Line Tools for Xcode 12.4".

If neither of those work, you can try running `brew doctor` to see if there are any other problems that Homebrew has detected. If you need help diagnosing any of those, feel free to post the output of `brew config` and `brew doctor` and we can try to help. Note, though, that the warnings that `brew doctor` shows don't always _need_ to be solved if everything is working okay.

---

To explain why `brew install cask` didn't do what you expected, you need to understand a little more about what Homebrew provides.

We allow for the installation of two types of packages. Some packages are called formulae. These are generally command-line tools that we pre-compile. The other type are called casks. Casks are generally used to install pre-built applications more easily. Essentially, it automates the normal process of downloading an application (downloading and opening a `.dmg` file, dragging and dropping the application into your `Applications` directory, running an installer, etc.). You can read more about the different terminology [here](https://docs.brew.sh/Formula-Cookbook#homebrew-terminology).

Both formula and cask functionality are built-in to Homebrew. You don't need to install anything (except Homebrew) to install casks. When you run `brew install foo`, you're telling Homebrew to install a package called `foo`. Homebrew will first check to see if there are any formulae called `foo` and, if not, if there are casks named `foo`. It will install the first one that it finds.

Here's where it gets a tad confusing: there is a formula called `cask`. I know, that's pretty unfortunate. When you ran `brew install cask`, you actually told Homebrew to "install a package called cask" and, because there's a formula with that name, it tried to install the formula. The same thing happened when you ran `brew info cask`. Homebrew displayed information about the formula called `cask`.

When you run `brew install foo`, Homebrew _will_ automatically install the dependencies for you. There's no need to look through to ensure that everything is already installed. We'll handle that for you. In case you are curious what dependencies are needed, you can run `brew deps --tree foo` to list the dependencies of `foo` in a tree-like format that's easy to understand.

So, in the future, if you want to install a cask, you can simply run `brew install foo` (where `foo` is the name of the cask you want to install) and Homebrew will take care of the rest for you.

---

To try to answer your question about terminal colors, you may want to check out the [`oh-my-zsh` themes](https://github.com/ohmyzsh/ohmyzsh/wiki/External-themes) and see if any of those work well for you. There are also instructions linked on that page explaining how to install/use the theme.

That's probably the easiest way to get colored output. If you really wanted/needed, you could create your own custom prompt, but this is a lot of work (especially for a beginner who isn't super familiar with the terminal and how that works).

---

If you want to get the hang of using the terminal, I would recommend checking out [CodeAcademy's Bash/Shell tutorial](https://www.codecademy.com/catalog/language/bash). It looks like they may have changed the website, though, and it may no longer be free.

I don't have any specific resources in mind, but I think you can find many tutorials if you do a little searching. One tutorial that came up that looks reasonable to me is https://ubuntu.com/tutorials/command-line-for-beginners. I only skimmed it, but it looks like a good resource to start. Note that while that tutorial is specifically about Linux, it should be totally applicable to macOS as well.

I hope that helps a little bit.

**Reply by COBRABYTE** [reply]
Thank you so much! I appreciate all the info you provided. And I'm also grateful for the extra advice you suggested regarding my bonus questions.

I think I resolved the errors that I was seeing. I did the brew doctor command and walked through the CLT issue, which then cleared up some of the other issues as well. It had me rm -rf the command line tools directory and then use the Xcode-select --install command to reinstall it. Once that finished and I ran brew doctor again, those issues were no longer there. 

After that, the remaining issues dealt with conflicting config files or something like that. Prior to installing Homebrew, I had installed Python manually from it's website. This had setup a python.frameworks directory I'm assuming as (maybe it was there before) but then within that directory it had a versions directory where it had 3.8 and 3.9 (I had originally downloaded 3.8 a while back, then recently downloaded 3.9). Having installed Homebrew after that I guess was causing conflicts due to how the $PATH was structured (I had no idea what $PATH was before yesterday, and I still don't fully understand what it is), but it seems to matter the order in which directories precede another (at least that is what I assume). After rearranging order of the stuff in the $PATH, nothing seemed to change (I still got those conflicting config issues with brew doctor). 

So then I completely removed python 3.8 and 3.9 from my applications folder as well as the corresponding python.framworks versions directories (3.8 and 3.9 and current [which seemed to be tied to either 3.8 or 3.9]). Then I used Homebrew to install python 3.9. Then on accident, meaning to do the man brew command for the brew manual, I typed brew man. I'm not really sure what this did, but afterwards when I ran brew doctor, I had no issues; it said I was ready to brew.

Also, thanks to you explaining cask and formulae to me, I figured out that I needed to use the command brew tap caskroom/cask. Before, when I had tried searching for casks, none ever came up. For example, I tried searching for Firefox, but nothing returned. Then when I searched for just fire, it only showed formulae that included fire in it. However, after doing the brew tap caskroom/cask command it seemed to resolve it. Now when I search things, I get results for both formulae and casks.

So at this point, my only question is, what did brew man do?

Also, bonus question:

For future reference, how do you insert commands into your post? For example when you referenced the commands that I did, they were in a distinctively different font and style from the rest of your response. I'd like to be able to do that as to make it easier on others when reading my posts, so that I don't have to constantly say command after it and such.

Again, thanks so much for taking the time to read and respond to my post! You're awesome!

**Reply by carlocab** [reply]
From `brew help`:
```
❯ brew help man
Usage: brew man [--fail-if-changed]

Generate Homebrew's manpages.
```

You don't really need to worry about `brew man` unless you make changes to `brew`'s codebase.

For code formatting, you can read about formatting markdown [here](https://guides.github.com/features/mastering-markdown/). (Look for the "Code" button under "Examples".) The buttons above your textbox where you enter text here might help too.

**Reply by COBRABYTE** [reply]
Yeah, I was just worried I might have accidentally screwed something up when I ran `brew man` instead of `man brew`, the latter of which I was actually intending. Once I realized something else was going on and the manual wasn't coming up, I was super worried. Anyways, thanks so much for your response. Also, thanks for pointing out the code text formatting.

**Reply by Rylan12** [reply]
The only thing I'd add to Carlo's answer is that `brew man` is a Homebrew developer command. These developer commands are designed to be run by those who are more familiar with Homebrew and, likely, are contributors. As a result, running this (or any) developer command acts as an opt-in to track the latest changes instead of only updating Homebrew when we release new versions. You can think of this as opting-in to become an alpha or beta tester.

My guess is that you don't want to opt-in here, so I'd recommend that you run the following command to opt-out:

```sh
git -C $(brew --repository) config --unset homebrew.devcmdrun
```

Then, run:

```sh
brew update
```
